src/core/mempool.rs

```diff
@@ -750,6 +777,198 @@ impl MemPoolTxInfo {
+                    Ok(opt_nonce) => opt_nonce,
+                    Err(e) => {
+                        warn!("error retrieving nonce from mempool db: {}", e);
+                        None
+                    }
+                };
+                match opt_nonce {
+                    Some(nonce) => {
+                        // Copy this into the in-memory cache if there is space
+                        if self.cache.len() < self.max_cache_size {
+                            self.cache.insert(address.clone(), nonce);
+                        }
+                        nonce
+                    }
+                    None => {
+                        let nonce =
+                            StacksChainState::get_nonce(clarity_tx, &address.clone().into());
+
+                        match db_set_nonce(mempool_db, address, nonce) {
+                            Ok(_) => (),
+                            Err(e) => warn!("error caching nonce to sqlite: {}", e),
+                        }
+
+                        if self.cache.len() < self.max_cache_size {
+                            self.cache.insert(address.clone(), nonce);
+                        }
+                        nonce
+                    }
+                }
+            }
+        }
+    }
+
+    fn update(&mut self, address: StacksAddress, value: u64, mempool_db: &DBConn) {
+        // Sqlite cache
+        match db_set_nonce(mempool_db, &address, value) {
+            Ok(_) => (),
+            Err(e) => warn!("error caching nonce to sqlite: {}", e),
+        }
+
+        // In-memory cache
+        match self.cache.get_mut(&address) {
+            Some(nonce) => {
+                *nonce = value;
+            }
+            None => (),
+        }
+    }
+}
+
+fn db_set_nonce(conn: &DBConn, address: &StacksAddress, nonce: u64) -> Result<(), db_error> {
+    let addr_str = address.to_string();
+    let nonce_i64 = u64_to_sql(nonce)?;
+
+    let sql = "INSERT OR REPLACE INTO nonces (address, nonce) VALUES (?1, ?2)";
+    conn.execute(sql, rusqlite::params![&addr_str, nonce_i64])?;
+    Ok(())
+}
+
+fn db_get_nonce(conn: &DBConn, address: &StacksAddress) -> Result<Option<u64>, db_error> {
+    let addr_str = address.to_string();
+
+    let sql = "SELECT nonce FROM nonces WHERE address = ?";
+    query_row(conn, sql, rusqlite::params![&addr_str])
+}
+
+/// Cache potential candidate transactions for subsequent iterations.
+/// While walking the mempool, transactions that have nonces that are too high
+/// to process yet (but could be processed in the future) are added to `next`.
+/// In the next pass, `next` is moved to `cache` and these transactions are
+/// checked before reading more from the mempool DB.
+struct CandidateCache {
+    cache: VecDeque<MemPoolTxInfoPartial>,
+    next: VecDeque<MemPoolTxInfoPartial>,
+    /// The maximum size that this cache can be.
+    max_cache_size: usize,
+}
+
+impl CandidateCache {
```

Comment:
Is `CandidateCache::cache` ever written to in this cache?  Does `CandidateCache::next()` ever return `Some(..)`?

Reply:
> Is CandidateCache::cache ever written to in this cache? 
`CandidateCache::cache` is only written to in `reset`.

> Does CandidateCache::next() ever return Some(..)?
Yes, see previous answer

Reply:
Ah, yes, I see that now.